folding-schemes/src/frontend/noname/mod.rs

```diff
@@ -0,0 +1,228 @@
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_relations::r1cs::{ConstraintSynthesizer, SynthesisError};
+use num_bigint::BigUint;
+use std::marker::PhantomData;
+
+use super::FCircuit;
+use ark_ff::PrimeField;
+use ark_noname::utils::compile_source_code;
+use noname::backends::{r1cs::R1CS as R1CSNoName, BackendField};
+use noname::witness::CompiledCircuit;
+
+pub mod utils;
+
+use utils::*;
+
+#[derive(Debug, Clone)]
+pub struct NoNameFCircuit<F: PrimeField, BF: BackendField> {
+    pub state_len: usize,
+    pub external_inputs_len: usize,
+    pub circuit: CompiledCircuit<R1CSNoName<BF>>,
+    _f: PhantomData<F>,
+}
+
+impl<F: PrimeField, BF: BackendField> FCircuit<F> for NoNameFCircuit<F, BF> {
+    type Params = (String, usize, usize);
+
+    fn new(params: Self::Params) -> Result<Self, crate::Error> {
+        let (code, state_len, external_inputs_len) = params;
+        let compiled_circuit = compile_source_code::<BF>(&code).map_err(|_| {
+            Error::Other("Encountered an error while compiling a noname circuit".to_owned())
+        })?;
+        Ok(NoNameFCircuit {
+            state_len,
+            external_inputs_len,
+            circuit: compiled_circuit,
+            _f: PhantomData,
+        })
+    }
+
+    fn state_len(&self) -> usize {
+        self.state_len
+    }
+
+    fn external_inputs_len(&self) -> usize {
+        self.external_inputs_len
+    }
+
+    fn step_native(
+        &self,
+        _i: usize,
+        z_i: Vec<F>,
+        external_inputs: Vec<F>,
+    ) -> Result<Vec<F>, crate::Error> {
+        // compute witness using noname witness generator and outputs the next step
+        let external_inputs_json = match !external_inputs.is_empty() {
+            false => None,
+            true => {
+                let json =
+                    serde_json::to_string(&NoNameJSONExternalInputs::from(external_inputs.clone()))
+                        .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+                Some(json)
+            }
+        };
+
+        let ivc_inputs = serde_json::to_string(&NoNameJSONIVCInput::from(z_i.clone()))
+            .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+        let noname_inputs = NoNameIVCInputs::<F>::try_from((ivc_inputs, external_inputs_json))
+            .map_err(|_| SynthesisError::Unsatisfiable)?;
+        let (ivc_inputs, noname_external_inputs) = noname_inputs.try_into()?;
+        let noname_witness = self
+            .circuit
+            .generate_witness(ivc_inputs, noname_external_inputs)
+            .map_err(|e| Error::WitnessCalculationError(e.to_string()))?;
+
+        let mut assigned_z_i1 = vec![];
+        let z_i1_end_index = z_i.len() + 1;
+        for idx in 1..z_i1_end_index {
+            let value: BigUint = Into::into(noname_witness.witness[idx]);
+            let field_element = F::from(value);
+            assigned_z_i1.push(field_element);
+        }
+
+        Ok(assigned_z_i1)
+    }
+
+    fn generate_step_constraints(
```

Comment:
As I always say in PRs. We need to make this fn as thin and fast as possible. So the less transformations and shenanigans we have on them the better.

Just saying to see if there's any way to simplify things.

Reply:
Agree, see the comment I did above, regarding the refactor of this PR